### PR TITLE
[#1026] Continuous Batched Enhancing

### PR DESCRIPTION
**Pull request by HonzaR**
- Tested with and analysed the related database entries with the Android Studio database inspector
- Measured its performance and it has no significant impact on the overall syncing process time
- Tested in parallel with the previous sync stages as well, but it does not give any significant improvement.
- Internal ENHANCING state eliminated as well, as described in https://github.com/zcash/zcash-android-wallet-sdk/issues/967
- We now treat the enhancing error as critical, as we do in case of other stages failures as well
- The enhancement range is changed in case of any previously un-enhanced transaction found in the transaction table view
- Changed enhancing trigger appropriately
- Covered also the case when the sync stage ended before the enhancement limit reached
- Cleaned up the code
- Closes https://github.com/zcash/zcash-android-wallet-sdk/issues/1026

> **Note**
>This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] **Self-review** your own code in GitHub's web interface[^1]
- [ ] Add **automated tests** as appropriate
- [ ] Update the [**manual tests**](../blob/main/docs/testing/manual_testing)[^2] as appropriate
- [ ] Check the **code coverage**[^3] report for the automated tests
- [ ] Update **documentation** as appropriate (e.g [README.md](../blob/main/README.md), [Architecture.md](../blob/main/docs/Architecture.md), etc.)
- [x] **Run the demo app** and try the changes
- [x] Pull in the latest changes from the **main** branch and **squash** your commits before assigning a reviewer[^4]

# Reviewer

- [x] Check the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) **checklist**
- [x] Perform an **ad hoc review**[^5]
- [ ] Review the **automated tests**
- [ ] Review the **manual tests**
- [x] Review the **documentation**, [README.md](../blob/main/README.md), [Architecture.md](/blob/main/docs/Architecture.md), etc. as appropriate
- [ ] **Run the demo app** and try the changes[^6]

[^1]: _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
[^2]: _While we aim for automated testing of the SDK, some aspects require manual testing. If you had to manually test 
something during development of this pull request, write those steps down._
[^3]: _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed. Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules._
[^4]: _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._
[^5]: _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
[^6]: _While the CI server runs the demo app to look for build failures or crashes, humans running the demo app are 
more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt [hunk 1]
```diff
@@ -1392,23 +1482,15 @@ class CompactBlockProcessor internal constructor(
...
      * block 11 but can't.
+     * @param firstUnEnhancedHeight the height in which the enhancing should start, or null in case of no previous
+     * transaction enhancing done yet
      */
     data class ProcessorInfo(
         val networkBlockHeight: BlockHeight?,
         val lastSyncedHeight: BlockHeight?,
-        val lastSyncRange: ClosedRange<BlockHeight>?
+        val lastSyncRange: ClosedRange<BlockHeight>?,
+        val firstUnEnhancedHeight: BlockHeight?
```
[the remark this answers is not in the document]
Comment (ccjernigan) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt]: I pushed a commit with this change.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt [hunk 2]
```diff
@@ -837,20 +788,61 @@ class CompactBlockProcessor internal constructor(
...
+                    // Increment and compare the range for triggering the enhancing
+                    enhancingRange = enhancingRange.start..continuousResult.batch.range.endInclusive
+
+                    // Enhance is run in case of the range is on or over its limit, or in case of any failure
+                    // state comes from the previous stages, or if the end of the sync range is reached
+                    if (enhancingRange.length() >= ENHANCE_BATCH_SIZE ||
+                        continuousResult.stageResult != BlockProcessingResult.Success ||
+                        continuousResult.batch.order == batches.size.toLong()
+                    ) {
+                        // Copy the range for use and reset for the next incrementation
```
[the remark this answers is not in the document]
Comment (ccjernigan) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt]: I pushed a commit with this change.

**Review by ccjernigan — APPROVED** [review]
This brings a lot of great improvements, both for performance and for robustness.  Thanks for the effort that went into this.

I have a few documentation questions, but those are non-blocking.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/Synchronizer.kt [hunk 3]
```diff
@@ -357,27 +357,27 @@ interface Synchronizer {
...
          * from the server. Blocks need to be verified before they are scanned. This confirms that
          * each block is chain-sequential, thereby detecting missing blocks and reorgs.
          *
-         * In **Scanning** sub-phase Synchronizer is actively decrypting new blocks that were downloaded
+         * In **Scanning** stage Synchronizer is actively decrypting new blocks that were downloaded
          * from the server.
+         *
+         * In **Deleting** stage are all temporary persisted block files removed from the persistence.
+         *
+         * In **Enhancing** stage is the Synchronizer actively enhancing newly scanned blocks with additional
```
[the remark this answers is not in the document]
Comment (HonzaR) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/Synchronizer.kt:376]: I understand your comment. I just made a few minor changes to clarify the documentation intention. All of the mentioned internal stages (downloading, validating, scanning, deleting, and enhancing) are removed now. We just keep this documentation comment to describe what is going on when is the Synchronizer in the SYNCING state. But we can omit them entirely as well.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt [hunk 4]
```diff
@@ -837,20 +788,61 @@ class CompactBlockProcessor internal constructor(
...
+                            Twig.debug { "Enhancing result: $enhancingResult" }
+                            when (enhancingResult) {
+                                is BlockProcessingResult.UpdateBirthday -> {
+                                    Twig.debug { "Birthday height update reporting" }
+                                }
+                                is BlockProcessingResult.FailedEnhance -> {
+                                    Twig.error { "Enhancing failed for: $enhancingRange with $enhancingResult" }
+                                }
+                                else -> {
+                                    // Transactions enhanced correctly
```
[the remark this answers is not in the document]
Comment (HonzaR) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt:829]: Yes, I was thinking about the same while I was on it. So I filed a follow-up for it https://github.com/zcash/zcash-android-wallet-sdk/issues/1047

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt [hunk 5]
```diff
@@ -1244,8 +1335,8 @@ class CompactBlockProcessor internal constructor(
     private fun calculatePollInterval(fastIntervalDesired: Boolean = false): Long {
```
[the remark this answers is not in the document]
Comment (HonzaR) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt:1335]: Yep, good idea. I just filed it here https://github.com/zcash/zcash-android-wallet-sdk/issues/1048

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (HonzaR) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt]: Thank you.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt [hunk 6]
```diff
@@ -1392,23 +1482,15 @@ class CompactBlockProcessor internal constructor(
      * @param lastSyncRange inclusive range to sync. Meaning, if the range is 10..10,
      * then we will download exactly block 10. If the range is 11..10, then we want to download
      * block 11 but can't.
+     * @param firstUnEnhancedHeight the height in which the enhancing should start, or null in case of no previous
```
[the remark this answers is not in the document]
Comment (HonzaR) [reply sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt]: Yes, the comment just updated.

**Review by ccjernigan — APPROVED** [review]

**Pull request merged**